state-chain/pallets/cf-ingress-egress/src/lib.rs

```diff
@@ -1641,33 +1647,105 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+					Some(ref refund_ccm) => IngressOrEgress::EgressCcm {
+						gas_budget: refund_ccm.channel_metadata.gas_budget,
+						message_length: refund_ccm.channel_metadata.message.len(),
+					},
+					None => IngressOrEgress::Egress,
+				},
 				tx.asset,
 				amount_after_ingress_fees,
 			);
-		if let Ok(api_call) = <T::ChainApiCall as RejectCall<T::TargetChain>>::new_unsigned(
+
+		// By building the ccm call, we will know if a separate "fetch" is required as part of the
+		// refund.
+		let no_ccm_refund = tx.refund_ccm_metadata.is_none();
+		match <T::ChainApiCall as RejectCall<T::TargetChain>>::new_unsigned(
 			tx.deposit_details.clone(),
-			refund_address,
-			amount_to_refund,
+			refund_address.clone(),
+			no_ccm_refund.then_some(amount_to_refund),
 			tx.asset,
 			deposit_fetch_id,
 		) {
-			let broadcast_id = T::Broadcaster::threshold_sign_and_broadcast_with_callback(
-				api_call,
-				tx.deposit_address.map(|deposit_address| {
-					Call::finalise_ingress { addresses: vec![deposit_address] }.into()
-				}),
-				|_| None,
-			);
-			Self::deposit_event(Event::<T, I>::TransactionRejectedByBroker {
-				broadcast_id,
-				tx_id: tx.deposit_details.clone(),
-			});
-		} else {
-			FailedRejections::<T, I>::append(tx.clone());
-			Self::deposit_event(Event::<T, I>::TransactionRejectionFailed {
-				tx_id: tx.deposit_details.clone(),
-			});
+			Ok(api_call) => {
+				let broadcast_id = T::Broadcaster::threshold_sign_and_broadcast_with_callback(
+					api_call,
+					tx.deposit_address.clone().map(|deposit_address| {
+						Call::finalise_ingress { addresses: vec![deposit_address] }.into()
+					}),
+					|_| None,
+				);
+				// Only emit TransactionRejectedByBroker for the actual egress (refund) transaction
+				if no_ccm_refund {
+					Self::deposit_event(Event::<T, I>::TransactionRejectedByBroker {
+						broadcast_id,
+						tx_id: tx.deposit_details.clone(),
+					});
+				}
+			},
+			Err(cf_chains::RejectError::NotRequired) => (),
+			Err(_) =>
+			// Only consider it failed if we are not doing a ccm refund
+				if no_ccm_refund {
+					FailedRejections::<T, I>::append(tx.clone());
+					Self::deposit_event(Event::<T, I>::TransactionRejectionFailed {
+						tx_id: tx.deposit_details.clone(),
+					});
+				},
+		};
+
+		if let Some(ref ccm_refund_metadata) = tx.refund_ccm_metadata {
+			let handle_ccm_failure = || {
+				FailedRejections::<T, I>::append(tx.clone());
+				Self::deposit_event(Event::<T, I>::TransactionRejectionFailed {
+					tx_id: tx.deposit_details.clone(),
+				});
+			};
+
+			// Solana CCM refunds with ALT is not supported, since we don't want to witness ALTs
+			// here.
+			if let DecodedCcmAdditionalData::Solana(sol_add_data) =
+				&ccm_refund_metadata.channel_metadata.ccm_additional_data
+			{
+				debug_assert!(
+					sol_add_data.alt_addresses().map(|alts| alts.is_empty()).unwrap_or(true),
+					"Solana refund CCM with ALTs is not supported"
+				);
+				handle_ccm_failure();
+				return;
```

Comment:
Do we consider this a failure even if the additional data is empty? 

(Should it not be `if !alt_addresses().is_empty() { handle_ccm_failure(); return; }` ?)